File: Days 01 - 10/Day 04/SmallestMissingPositive.cpp
#include <iostream>
#include <utility>
#include <vector>

unsigned int SplitPositives(std::vector<int>& numbers)
{
	unsigned int nonPositiveCount = 0;

	for (unsigned int i = 0; i < numbers.size(); i++)
	{
		if (numbers[i] <= 0)
		{
			std::swap(numbers[i], numbers[nonPositiveCount]);

			nonPositiveCount++;
		}
	}

	return nonPositiveCount;
}

unsigned int GetSmallestMissingPositive(std::vector<int>& numbers)
{
	unsigned int nonPositiveCount = SplitPositives(numbers);

	if (nonPositiveCount == numbers.size())
	{
		return 1;
	}

	unsigned int positiveCount = numbers.size() - nonPositiveCount;

	for (unsigned int i = nonPositiveCount; i < numbers.size(); i++)
	{
		if (numbers[i] - 1 < positiveCount)
		{
			numbers[numbers[i] - 1] *= -1;
		}
	}

	for (unsigned int i = nonPositiveCount; i < numbers.size(); i++)
	{
		if (numbers[i] > 0)
		{
			return i + 1;
		}
	}

	return positiveCount + 1;
}

int main(int argc, char* argv[])
{
	std::vector<int> numbers { 3, 4, -1, 1 };
	std::cout << GetSmallestMissingPositive(numbers) << "\n";

	std::vector<int> moreNumbers { 1, 2, 0 };
	std::cout << GetSmallestMissingPositive(moreNumbers) << "\n";

	std::cin.get();
	
	return 0;
}

File: Days 01 - 10/Day 04/SmallestMissingPositive.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblems
{
	internal class Day4
	{
		private static int Main(string[] args)
		{
			List<int> numbers = new List<int> { 3, 4, -1, 1 };
			Console.WriteLine(GetSmallestMissingPositive(numbers));

			numbers = new List<int> { 1, 2, 0 };
			Console.WriteLine(GetSmallestMissingPositive(numbers));

			Console.ReadLine();

			return 0;
		}

		private static int GetSmallestMissingPositive(List<int> numbers)
		{
			int nonPositiveCount = SplitPositives(numbers);

			if (nonPositiveCount == numbers.Count)
			{
				return 1;
			}

			int positiveCount = numbers.Count - nonPositiveCount;

			for (int i = nonPositiveCount; i < numbers.Count; i++)
			{
				if (numbers[i] - 1 < positiveCount)
				{
					numbers[numbers[i] - 1] *= -1;
				}
			}

			for (int i = nonPositiveCount; i < numbers.Count; i++)
			{
				if (numbers[i] > 0)
				{
					return i + 1;
				}
			}

			return positiveCount + 1;
		}

		private static int SplitPositives(List<int> numbers)
		{
			int nonPositiveCount = 0;

			for (int i = 0; i < numbers.Count; i++)
			{
				if (numbers[i] <= 0)
				{
					int temp = numbers[i];
					numbers[i] = numbers[nonPositiveCount];
					numbers[nonPositiveCount] = temp;

					nonPositiveCount++;
				}
			}

			return nonPositiveCount;
		}
	}
}